src/ChunkedStreamDecoder.php

```diff
@@ -0,0 +1,147 @@
+<?php
+
+namespace React\HttpClient;
+
+use Evenement\EventEmitterTrait;
+use Exception;
+use React\Stream\ReadableStreamInterface;
+use React\Stream\Util;
+use React\Stream\WritableStreamInterface;
+
+class ChunkedStreamDecoder implements ReadableStreamInterface
+{
+    const CRLF = "\r\n";
+
+    use EventEmitterTrait;
+
+    /**
+     * @var string
+     */
+    protected $buffer = '';
+
+    /**
+     * @var int
+     */
+    protected $remainingLength = 0;
+
+    /**
+     * @var bool
+     */
+    protected $nextChunkIsLength = true;
+
+    /**
+     * @var ReadableStreamInterface
+     */
+    protected $stream;
+
+    /**
+     * @param ReadableStreamInterface $stream
+     */
+    public function __construct(ReadableStreamInterface $stream)
+    {
+        $this->stream = $stream;
+        $this->stream->on('data', array($this, 'handleData'));
+        Util::forwardEvents($this->stream, $this, [
+            'error',
+        ]);
+    }
+
+    /** @internal */
+    public function handleData($data)
+    {
+        $this->buffer .= $data;
+
+        do {
+            $bufferLength = strlen($this->buffer);
+            $continue = $this->iterateBuffer();
+            $iteratedBufferLength = strlen($this->buffer);
+        } while (
+            $continue &&
+            $bufferLength !== $iteratedBufferLength &&
+            $iteratedBufferLength > 0 &&
+            strpos($this->buffer, static::CRLF) !== false
+        );
+    }
+
+    protected function iterateBuffer()
+    {
+        if ($this->nextChunkIsLength) {
+            $crlfPosition = strpos($this->buffer, static::CRLF);
+            if ($crlfPosition === false) {
+                return false; // Chunk header hasn't completely come in yet
+            }
```

Comment:
This can be used for a DOS because all data will be buffered until a CRLF has been found (see also above).

We should probably apply a limit here (1K or so) and then emit an error if no CRLF can be found.